controller/state.go

```diff
@@ -305,6 +309,58 @@ func verifyGnuPGSignature(revision string, project *appv1.AppProject, manifestIn
 	return conditions
 }
 
+func (m *appStateManager) diffArrayCached(configArray, liveArray []*unstructured.Unstructured, opts ...diff.Option) (*diff.DiffResultList, error) {
+	numItems := len(configArray)
+	if len(liveArray) != numItems {
+		return nil, fmt.Errorf("left and right arrays have mismatched lengths")
+	}
+
+	managedResources := make([]*appv1.ResourceDiff, 0)
+	if m.cache.GetAppManagedResources("", &managedResources) != nil {
+		// (rare) cache miss
+		return diff.DiffArray(configArray, liveArray, opts...)
+	}
+
+	diffResultList := diff.DiffResultList{
+		Diffs: make([]diff.DiffResult, numItems),
+	}
+
+	resourceMap := make(map[kube.ResourceKey]*unstructured.Unstructured)
+
+	for i := 0; i < numItems; i++ {
+		key := kube.GetResourceKey(configArray[i])
+		resourceMap[key] = configArray[i]
+	}
+
+	for i := 0; i < numItems; i++ {
+		var dr *diff.DiffResult
+		if i < len(managedResources) {
+			cachedResource := managedResources[i]
+			key := kube.NewResourceKey(cachedResource.Group, cachedResource.Kind, cachedResource.Namespace, cachedResource.Name)
+			targetObj := resourceMap[key]
+			if cachedResource.ResourceVersion == targetObj.GetResourceVersion() {
+				dr = &diff.DiffResult{
+					NormalizedLive: []byte(cachedResource.NormalizedLiveState),
+					PredictedLive:  []byte(cachedResource.PredictedLiveState),
+					Modified:       cachedResource.NormalizedLiveState == cachedResource.PredictedLiveState,
```

Comment:
`NormalizedLiveState` and `PredictedLiveState` are always different. Only way to avoid re-comparing is to store "Modified" in the cache.  Recently I've proposed to store `SyncStatus` field but now I see that we can just store "Modified bool"